### PR TITLE
Speed up range scans with range tombstones

### PR DESCRIPTION
**Pull request by abhimadan**
Summary: Previously, every range tombstone iterator was seeked on every
ShouldDelete call, which quickly degraded performance for long range
scans. This PR improves performance by tracking iterator positions and
only advancing iterators when necessary.

Test Plan: make check

### CODE REVIEW

**Comment by abhimadan** [comment]
Removed WIP but there's still quite a bit of code duplication that I'm not too fond of, and there's some iterator cleanup logic that I want to try out. The general approach is here though.

File: db/range_del_aggregator_v2.h [hunk 1]
```diff
@@ -78,6 +78,175 @@ class TruncatedRangeDelIterator {
...
+  }
+
+  TruncatedRangeDelIterator* PopInactiveIter() {
+    auto* iter = inactive_iters_.top();
+    inactive_iters_.pop();
+    return iter;
+  }
+
+  void PushConsumedIter(TruncatedRangeDelIterator* iter) {
+    consumed_iters_.push_back(iter);
```
Comment (ajkr) [inline db/range_del_aggregator_v2.h]: I forget why do we keep `consumed_iters_`?

File: db/range_del_aggregator_v2.h [hunk 2]
```diff
@@ -78,6 +78,175 @@ class TruncatedRangeDelIterator {
...
+  ReverseRangeDelIterator(
+      const InternalKeyComparator* icmp,
+      const std::vector<std::unique_ptr<TruncatedRangeDelIterator>>* iters);
+
+  bool ShouldDelete(const ParsedInternalKey& parsed);
+  void Invalidate();
+
+ private:
+  using ActiveSeqSet =
+      std::multiset<TruncatedRangeDelIterator*, SeqMaxComparator>;
```
Comment (ajkr) [inline db/range_del_aggregator_v2.h]: in what case do we need `std::multiset` over `std::set`?

Comment (ajkr) [reply db/range_del_aggregator_v2.h]: oh never mind, I understood now it is storing multiple elements for a given seqnum, not multiple elements for a given pointer .

File: db/range_del_aggregator_v2.cc [hunk 3]
```diff
@@ -104,9 +104,124 @@ void TruncatedRangeDelIterator::SeekToFirst() { iter_->SeekToTopFirst(); }
...
+
+  // Move inactive iterators that end after parsed.
+  while (!inactive_iters_.empty() &&
+         icmp_->Compare(parsed, inactive_iters_.top()->end_key()) < 0) {
+    TruncatedRangeDelIterator* iter = PopInactiveIter();
+    while (iter->Valid() && icmp_->Compare(parsed, iter->start_key()) < 0) {
+      iter->Prev();
+    }
+    PushIter(iter, parsed);
+    assert(active_iters_.empty() || !active_seqnums_.empty());
```
Comment (ajkr) [inline db/range_del_aggregator_v2.cc]: should the size of these collections be equal?

File: db/range_del_aggregator_v2.h (hunk 1 above)
Comment (abhimadan) [reply db/range_del_aggregator_v2.h]: We actually don't need it so I'll remove it; thanks for noticing this.

File: db/range_del_aggregator_v2.cc (hunk 3 above)
Comment (abhimadan) [reply db/range_del_aggregator_v2.cc]: They should, but `BinaryHeap`s don't expose a `size()` method. I didn't want to add too much debug code for this invariant so I thought this was a decent compromise, though I'm open to adding a more sophisticated check here.

**Pull request closed without merging**